File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> best_buy(prices.size());
        vector<int> best_sell(prices.size());

        int n = prices.size();
        best_buy[0] = prices[0];
        best_sell[n - 1] = prices[n - 1];

        for (int i = 1; i < n; i++) {
            best_buy[i] = min(best_buy[i - 1], prices[i]);
        }
        for (int i = n - 2; i >= 0; i--) {
            best_sell[i] = max(best_sell[i + 1], prices[i]);
        }

        int max_profit = 0;
        for (int i = 0; i < n; i++) {
            max_profit = max(best_sell[i] - best_buy[i], max_profit);
        }

        return max_profit;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        max_area = -1
        n = len(heights)

        left = 0
        right = n - 1

        while left < right:
            breadth = right - left
            height = min(heights[left], heights[right])
            area = breadth * height
            max_area = max(area, max_area)

            if (heights[left] <= heights[right]):
                left += 1
            elif (heights[left] > heights[right]):
                right -= 1
        
        return max_area
